test/distributed/org/apache/cassandra/distributed/test/HintDataReappearingTest.java

```diff
@@ -0,0 +1,267 @@
+                    .drop();
+
+
+            logger.info("Pausing hint delivery ");
+            // pause hint delivery to imitate hints being behind/backed up
+            pauseHintsDelivery(node1);
+
+            logger.info("Inserting data");
+
+            List<UUID> keys = IntStream.range(0, 1).mapToObj(x -> UUID.randomUUID()).collect(Collectors.toList());
+
+            scheduler.submit(() -> {
+                if (dropTwoWrites)
+                {
+                    Assertions.assertThatThrownBy(() -> insertData(cluster, keys))
+                            .is(isThrowable(WriteTimeoutException.class));
+                }
+                else
+                {
+                    insertData(cluster, keys);
+                }
+            });
+
+            Thread.sleep(1000);
+            dropWritesForNode2.set(false);
+            dropWritesForNode3.set(false);
+
+            node1.flush(KEYSPACE);
+            node2.flush(KEYSPACE);
+            node3.flush(KEYSPACE);
+
+            logger.info("deleting data ");
+            deleteData(cluster, keys);
+            long afterDelete = System.currentTimeMillis();
+
+            node1.flush(KEYSPACE);
+            node2.flush(KEYSPACE);
+            node3.flush(KEYSPACE);
+
+
+            logger.info("repairing");
+            for (IInvokableInstance node : Arrays.asList(node1, node2, node3))
+            {
+                node.nodetoolResult(ArrayUtils.addAll(new String[]{ "repair", KEYSPACE }, "--full")).asserts().success();
+            }
+            logger.info("done repairing");
+
+
+            for (SimpleQueryResult result : selectData(cluster, keys))
+            {
+                Object[][] objectArrays = result.toObjectArrays();
+                logger.info("result after delete: {} {}", result, Arrays.deepToString(objectArrays));
+                // We expect the data to appear to be deleted initially
+                Assert.assertNull(objectArrays[0][1]);
+            }
+
+            // wait to pass gc_grace_seconds with slight buffer of 2 seconds to ensure delete persisted long enough to be gced
+            long msSinceDelete = Math.abs(System.currentTimeMillis() - afterDelete);
+            long sleepFor = Math.max(0, 1000 * gc_grace_seconds + 2000 - msSinceDelete);
+            logger.info("sleeping {} ms to ensure gc_grace_seconds has ellapsed after tombstone creation", sleepFor);
+            Thread.sleep(sleepFor);
+
+            // ensure tombstone purged on all 3 nodes
+            node1.forceCompact(KEYSPACE, "cf");
+            node3.forceCompact(KEYSPACE, "cf");
+            node2.forceCompact(KEYSPACE, "cf");
+
+
+            IIsolatedExecutor.CallableNoExcept<UUID> node2hostId = node2.callsOnInstance(SystemKeyspace::getLocalHostId);
+
+            transferHints(node1, node2hostId.call());
+
+            // Sleep a bit more to ensure hint is delivered after tombstone is GCed
+            Thread.sleep(20000);
```

Comment:
20s is super long and gonna make this test a bit more of a pain, think we can slim this down some?